### ISSUE TITLE
Raise minimum version of PHP to 7.2 for 4.0

### ISSUE THREAD

**Issue by markstory** [cakephp/cakephp#13103]
We are currently planning to release CakePHP 4.0 requiring at least PHP 7.1. However, PHP 7.1 will become [end-of-life at the end of this year](https://www.php.net/supported-versions.php). Given that we are about to start the release process for 4.0, we have the opportunity to raise the minimum PHP version to 7.2.

We'd like to gather feedback from the community on which version we should require. 

* Add a :+1: to vote for PHP 7.2
* Add a :-1: to vote for PHP 7.1

**Comment by dmcdaniel12** [comment]
Ideally it would be 7.3 but 7.2 EOL isn't until next year so plenty of time to have this discussion again in a few months :)

**Comment by Bird87ZA** [comment]
Was literally about to post what @dmcdaniel12 posted. My feeling - even though I understand that EOL isn't until next year - is that new major releases should require newest available versions where possible. Since the step up between 7.2 and 7.3 isn't huge, I feel like you might just as well require 7.3

**Comment by AshleyPinner** [comment]
I wouldn't require 7.3 just yet as getting it in even 3rd party repos isn't quite there yet (webtatic as an example stops at 7.2). If the step between 7.2 and 7.3 isn't huge then there's no reason to require 7.3 either, the sword cuts both ways.

If you can argue that there's a feature of 7.3 that Cake4 could make great use of then the discussion might be a bit different, but given the current facts of 7.1's impending EoL, requiring 7.2 is a no-brainer. Moving that requirement to 7.3 is absurd at this point though with 7.2 not being EOL for a long while yet.

**Comment by Bird87ZA** [comment]
This is true. As you say, it cuts both ways. And I didn't think about 3rd party repos (amateur, I know)

Let's then have this discussion again in a few months.

**Comment by ADmad** [comment]
Just being the devils advocate here: The decision about what min. PHP version framework should support should be based on the benefits the core and app code can achieve by bumping to higher version, not just EOL of PHP version. We know that in the *real* world there are always systems which are forced to continue using older PHP version for X reason.

Checking the [new](https://www.php.net/manual/en/migration72.new-features.php) feature of PHP 7.2, off top of my head I don't see any being particularly helpful in core code. The new `object` type hinting and parameter widening would still be usable in app code if one deploys on PHP 7.2.

Of course people should not deploy their apps on PHP versions which are nearing or have reaching EOL but as long Cake v4 supports PHP 7.2/7.3 having 7.1 as the min. supported version doesn't matter.

**Comment by dereuromark** [comment]
I agree that racing to extremes for no reason is not helpful to anyone. Finding a reasonable balance of letting servers getting updated to the minimum and not having to support too many (legacy) versions is the correct approach IMO.
As such both 7.1 and 7.2 at this time are OK, with Cake4 still needing a bit of time, 7.2 could be the reasonable choice soon enough. So all good.

**Comment by inoas** [comment]
Didn't the major PHP version set for the major framework version release seems to stay especially internally for a long time? Even if the minimum version is upped due to no support / security / new features required, the old code stays the same unless there is an compatibility issue. So the style the framework exposes does not upgrade.

As 4.x will be there for quite a time and 3.x is quite mature those who need to stick to <7.3 can do so picking 3.8.

So I am glad if you choose 7.2 but I don't see why choosing 7.3 now is an issue for new projects running on 4.x?

What's new in 7.3: https://kinsta.com/blog/php-7-3/

This I find note-able in 7.3:
- Trailing commas in function calls
- Heredoc/Nowdoc syntax
- json_encode throwing Execptions
- CompileError Exception instead of RuntimeError Exception in a few cases
- array_key_first and array_key_last over mutating arrays with end and reset.
- compact() being strict https://www.php.net/manual/en/function.compact.php

**Comment by dereuromark** [comment]
> I don't see why choosing 7.3 now is an issue for new projects

Cake4.1 can still go 7.3, Cake4.2 could go 7.4 and so one. Nothing prevents progress in minors.
But killing support for many server solutions out there prematurely just for this is worse usually. That's not responsible framework development.

**Comment by ADmad** [comment]
Cake 4.0 won't be jumping to PHP 7.3, so guys please refrain from advocating for it and adding noise to this discussion :slightly_smiling_face:. As @dereuromark mentioned we can always revisit the min. version for Cake 4.1.

**Comment by Spriz** [comment]
While I understand the pressure to move to new PHP versions fast, I think the reality "out there" is quite different.

We're still running PHP 5.6 in production - and while we're working on moving to PHP 7.3 I know lots of companies do not have that option *right now*. I don't think we should exclude users from Cake 4.x because they're on 7.1 unless there is some features in 7.2 that are really worth the sacrifice.

TLDR: Only change requirements to 7.2 if there is features that are mush-have in 7.2 for framework-code. App code can still use 7.2-only features

**Comment by stripthis** [comment]
IMHO it would be useful if the PHP version aligns with the version supported on major Linux distros since it's very likely a lot of users will deploy apps on the distro their hosting provider is giving them.

The current Debian stable is still at 7.0 (with the debian team doing security fixes for a year or more) and the next Debian stable is targeting 7.3 (probably available mid-summer). The Ubuntu LTS is on 7.2. I think supporting 7.2 wouldn't hurt considering this landscape.

**Comment by ADmad** [comment]
@markstory Yeah adding typehints was tough as some places but we have already done the hard work. How do you see parameter widening helping us now in core?

**Comment by markstory** [comment]
@Spriz I hear you on the difficulties of keeping php updated.

From what we've seen in 2.x users hey are for more willing and able to update php than cakephp.

Checking what the current LTS releases of redhat, debian, and ubuntu come packaged with is a good idea.

**Comment by josegonzalez** [comment]
- Ubuntu packages 7.2 in their LTS release: https://packages.ubuntu.com/bionic/php
- Debian Stretch (stable) is 7.0: https://packages.debian.org/stable/php/php
- Debian Buster (testing) is 7.3: https://packages.debian.org/buster/php
  - This will likely be stable sometime in mid-2019
- Centos 7/Redhat apparently have 5.4: https://distrowatch.com/table.php?distribution=centos

A nicer list is available here: http://phpversions.info/operating-systems/

As a point of reference - not to validate/invalidate any options - what do other major PHP frameworks have as minimum versions? As well, it would be nice to see what our Japanese core/community members think, as PHP version lagging on VPS hosts in Japan has been an issue there in the past...

**Comment by nojimage** [comment]
Redhat 8.0(beta) is PHP 7.2

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8-beta/html/8.0_beta_release_notes/new-features#web_servers_databases_dynamic_languages_2

**Comment by voycey** [comment]
Yeah I would agree with the above - LTS versions of Ubuntu install 7.2 by default and seeing as so many use this as a base to develop / host it gets my vote!

**Comment by myllymakim** [comment]
If there are no PHP 7.2 features needed, I agree that PHP EOL is not best idea to bump up to PHP 7.2. requirement. Many systems have backporting and they still get PHP security updates. Still people complain that they have old and unsecure PHP. They do not understand stable enterprise systems.

**Comment by alexdd55** [comment]
@myllymakim Our admin has the same opinion as you have, but also sees that we are not able to meet requirements of a high quality product (incl. testing, MongoDB, encryption types, etc.) with the current Debian version (5.6.x). With upgrading our quality will rise all over the product. 
If this is not neccessary, then there is also no need to upgrade your Cake Version and argue about it. ;)

**Comment by ulistaerk** [comment]
PHP 7.2 is included in the LTS release of Ubuntu 18.04 and RHEL-8! I think this is feasible for Sysadmins and see no reason not to move forward.

**Comment by markstory** [comment]
Thanks for all the feedback on this everyone. I've opened a pull request to bump the minimum version of PHP up to 7.2.
